Add rpm for trino-cli

hashhar opened the pull request:
Adds the rpm-maven-plugin to the package phase to create an RPM that
will install the presto-cli executable jar as presto-cli to
/usr/local/bin. Fixes #1882.

hashhar commented:
This is WIP for a pure Java implementation of https://github.com/prestosql/presto/pull/3958 that doesn't depend on any external commands or a different build environment.

hashhar commented on pom.xml:
```diff
@@ -121,6 +121,8 @@
         <module>presto-product-tests-launcher</module>
         <module>presto-jdbc</module>
         <module>presto-cli</module>
+        <module>presto-cli-provisio</module>
```
Does the module name look okay? Presto server was broken down as:

`presto-server` - provisio assembly descriptor, since `presto-cli` was already taken I chose `presto-cli-provisio`.
`presto-server-main` - actual presto server module equivalent to `presto-cli`
`presto-server-rpm` - rpm module for presto server equivalent to `presto-cli-rpm`

kokosing replied:
Can we have presto-cli-main (the current presto-cli renamed), presto-cli that would produce presto-cli-executable and provisio package and then presto-cli-rpm? That way we will be coherent with presto-server. WDYT?

hashhar replied:
It will be coherent but since it was an invasive/breaking change (people expect presto-cli to contain code today) I did not feel confident enough to take that decision.
If you feel it is ok, I can re-arrange accordingly.

findepi replied:
Does it mean artifact names for presto-cli would change? eg links to maven central?

hashhar replied:
@findepi Yes. The artifactId will change.

findepi replied:
> The artifactId will change.

you mean for CLI code, or for final executable?

hashhar replied:
The existing executable JAR that gets created will change and the CLI code. The RPM artifact will remain the same.

findepi replied:
> The existing executable JAR that gets created will change

That's a bummer. Can we avoid that?

hashhar replied:
I don't think so. Only the suffix is configurable. The artifact name will always contain the `artifactId` AFAIK. Maybe @jvanzyl might have some ideas.

hashhar replied:
Another option is to rename `presto-cli-provisio` --> `presto-cli-assembly` since that will at-least make the intent clear and be familiar to people from other projects.
Unfortunately renaming `presto-cli` would definitely mean breaking any scripts that download the artifact from Maven central.

findepi replied:
i don't want this inline conversation to 'disappear', so -> https://github.com/prestosql/presto/pull/4025#issuecomment-650175037

sopel39 replied:
Maybe if building cli RPM takes few seconds we could simply inline all modules into one? Is that even possible?

hashhar replied:
Sadly that doesn't work due to reasons outlined in [here](https://github.com/prestosql/presto/pull/4025/files#diff-4ba198ba64f21a442f8a95c010e07ffbR2-R6). I'd really like for that comment being proven false though. :cry:

sopel39 replied:
For me it's fine to have:
* `presto-cli-assembly`
* `presto-cli-rpm`
* `presto-cli`
but we should add a comment why this extra modules are needed in their `pom.xml`s

hashhar commented on pom.xml:
```diff
@@ -1326,6 +1328,12 @@
                     <version>1.0.9</version>
                 </plugin>
+
+                <plugin>
```
Pulled out to ensure common versions are used everywhere (`presto-server-rpm` and `presto-cli-rpm`).

hashhar commented on presto-cli-provisio/src/main/provisio/presto.xml:
```diff
@@ -0,0 +1,10 @@
+    <archive name="${project.artifactId}-${project.version}.tar.gz" />
+
+    <!-- Presto CLI executable JAR -->
+    <!-- TODO: This should not add dependency jars into the archive -->
```
@jvanzyl Is there any way to tell provisio to only include the exact artifact specified without pulling in any dependencies?

hashhar commented on presto-cli-rpm/pom.xml:
```diff
@@ -0,0 +1,165 @@
+
+    <profiles>
+        <profile>
+            <id>ci</id>
```
The IT is added only to CI profile to mirror the `presto-server-rpm` module. The tests currently take around 10s from start to finish.

electrum replied:
Let's remove this so that it always runs. The reason `presto-server-rpm` has it was due to how the CI used to work when we ran with Travis CI, so we should be able to remove it from there as well.

hashhar commented:
@findepi @kokosing This is ready for review. Targets `337-SNAPSHOT`.

yuokada commented on presto-cli-rpm/pom.xml:
```diff
@@ -0,0 +1,165 @@
+
+                            <group>Applications/Databases</group>
+                            <description>Presto CLI RPM Package.</description>
+                            <architecture>x86_64</architecture>
```
Presto support both x86_64 and aarch64. presto-cli should support both, too.
Fortunately, as far as I know, presto-cli.jar doesn't have dependencies to CPU architectures.
I assume we can select `noarch` inatead.

hashhar replied:
Good point. I'll test on an aarch64 system with the changes and update accordingly.

hashhar replied:
It works. Updated to noarch.

findepi commented:
We should not change `presto-cli-${version}-executable.jar` name in central.
What other assumptions do we have? if this is the only one, we should be able to pick the names appropriately.

I am aware this _**may**_ require moving all the cli code around to e.g. `presto-cli-main` new module.

Since moving code around is time consuming, i suggest:

1. agree on the overall structure without any code
2. do a move PR which just moves the code, without adding new RPM module
3. do a PR adding CLI rpm (ie this one).

@hashhar @sopel39 @electrum @kokosing  WDYT?

hashhar commented:
@findepi The challenge is that as soon we rename an existing module it's maven co-ordinates will change and hence the Maven central URL. This makes renaming `presto-cli` safely impossible.

That leaves us with an easier problem to solve (or maybe one of the two most difficult problems in computer science) - deciding the name for `presto-cli-provisio`. I propose to rename it as `presto-cli-assembly` since it conveys what it's supposed to be.

I dug into the 333 -> 334 transition and saw that this issue did not arise when `presto-server-main` was added since no module renaming was done. `presto-server` has always contained the provisio assembly and `presto-server-rpm` has always been the RPM. `presto-server-main` was a completely new module (added to assist with tests???) and hence no issues.

sopel39 commented on presto-cli-rpm/pom.xml:
```diff
@@ -0,0 +1,165 @@
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-failsafe-plugin</artifactId>
+                        <version>2.22.2</version>
```
Please extract `maven-failsafe-plugin` version to `pom.xml` and use that version here and  `presto-server-rpm`

hashhar replied:
Done.

sopel39 commented on presto-cli-provisio/pom.xml:
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
please add a comment what this module is supposed to do and why we need a different module apart from `presto-cli-rpm`

hashhar replied:
It's a little long but [here](https://github.com/prestosql/presto/pull/4025/files#diff-4ba198ba64f21a442f8a95c010e07ffbR2-R6).

sopel39 replied:
Could you make this part of `pom.xml`, so that future code readers can get the idea?

hashhar replied:
The root pom.xml? Or this pom? I've added it to `presto-cli-provisio/pom.xml`.

sopel39 commented on presto-cli-rpm/pom.xml:
```diff
@@ -0,0 +1,165 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
How long does it take to run `presto-cli-rpm` and `presto-cli-provisio`?

hashhar replied:
<10 seconds for provisio and ~2m for presto-cli-rpm with tests (~10 seconds without tests).

hashhar commented:
@sopel39 Just to be on the same page, the requested changes are:

1. Rename `presto-cli-provisio` -> `presto-cli-assembly`.
2. Add a comment in `pom.xml` of EACH `presto-cli-rpm` and `presto-cli-assembly` documented why the 2 new modules were added.

Is my understanding correct?

hashhar commented:
Squashed commits, renamed `presto-cli-provisio` -> `presto-cli-assembly`, added comments to pom.xml of new modules explaining what each is doing and why it has to be done this way. Also updated the commit message to reflect the changes.

sopel39 commented:
> Is my understanding correct?

Yup

electrum commented:
I found a way to make the RPM build work in the main `presto-cli` module -- just copy the file into the `classes` directory. This is silly, but seems to work fine:
```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-antrun-plugin</artifactId>
    <version>3.0.0</version>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>run</goal>
            </goals>
            <configuration>
                <target>
                    <copy todir="${project.build.outputDirectory}" overwrite="true">
                        <fileset dir="${project.build.directory}">
                            <include name="presto-cli-${project.version}-executable.jar" />
                        </fileset>
                    </copy>
                </target>
            </configuration>
        </execution>
    </executions>
</plugin>
```
Note that we use our own fork of the RPM plugin, so we could probably also fix it if needed: https://github.com/airlift/redlinerpm-maven-plugin

hashhar commented:
@electrum I've incorporated all your comments except the antrun invocation. That looks like fighting against Maven and I'm not sure how future proof/safe it is.

Thanks a lot for the solution that we can directory copy jars using provisio. It allowed me to now remove the `presto-cli-assembly` module.

hashhar commented:
@electrum Could you take a look at https://github.com/prestosql/presto/pull/4025#issuecomment-651381756 and provide inputs whenever you get the chance?

electrum commented:
Using `antrun` is fine. We are simply working around a limitation in the RPM plugin. It’s probably easy to fix there, so I’m happy to review a patch if you want to try that instead: https://github.com/airlift/redlinerpm-maven-plugin

electrum commented:
Please also remove the `ci` profile.

hashhar commented:
Really sorry for how messed up this entire implementation went. 😞

I ended up getting in touch with the original author of the redlinerpm plugin and it turns out that there is a `buildPath` configuration for the plugin which can be set to whatever we need. It's only used for specifying the directory to be scanned and doesn't write any files there (unless it also happens to be the `target` directory). Was able to get rid of antrun too.

Thanks everyone for the patience.

hashhar commented:
@electrum Requesting a re-review.

hashhar commented:
@findepi No flags to skip RPM creation since the plugin doesn't implement a `<skip>` configuration. It's associated with the `package` phase so anything that skips `package` will skip RPM creation. The RPM creation is quick by the way. The time taking part is the RPM integration tests which can be skipped similar to other tests.

The time taken by the quick build command in your gist is ~4m40s both before and after this change.

Rebased against latest master and pushed.
